das/state.go

```diff
@@ -1,53 +1,228 @@
-	sample   RoutineState // tracks information related to the main sampling routine
+// coordinatorState represents the current state of sampling
+type coordinatorState struct {
+	rangeSize uint64
 
-	catchUpLk sync.RWMutex
-	catchUp   JobInfo // tracks information related to the `catchUp` routine
+	priority   []job                      // list of headers heights that will be sampled with higher priority
+	inProgress map[int]func() workerState // keeps track of running workers
+	failed     map[uint64]int             // stores heights of failed headers with amount of attempt as value
+
+	nextJobID   int
+	next        uint64 // all headers before next were sent to workers
+	networkHead uint64
+
+	catchUpDone   bool          // indicates if all headers are sampled
+	catchUpDoneCh chan struct{} // blocks until all headers are sampled
 }
 
-// RoutineState contains important information about the state of a
-// current sampling routine.
-type RoutineState struct {
-	// reports if an error has occurred during the routine's
-	// sampling process
-	Error error `json:"error"`
-	// tracks the latest successfully sampled height of the routine
-	LatestSampledHeight uint64 `json:"latest_sampled_height"`
-	// tracks the square width of the latest successfully sampled
-	// height of the routine
-	LatestSampledSquareWidth uint64 `json:"latest_sampled_square_width"`
-	// tracks whether routine is running
-	IsRunning bool `json:"is_running"`
+// newCoordinatorState initiates state for samplingCoordinator
+func newCoordinatorState(samplingRangeSize uint64) coordinatorState {
+	return coordinatorState{
+		rangeSize:     samplingRangeSize,
+		priority:      make([]job, 0),
+		inProgress:    make(map[int]func() workerState),
+		failed:        make(map[uint64]int),
+		nextJobID:     0,
+		next:          genesisHeight,
+		networkHead:   genesisHeight,
+		catchUpDone:   false,
+		catchUpDoneCh: make(chan struct{}),
+	}
+}
+
+func (s *coordinatorState) resumeFromCheckpoint(c checkpoint) {
+	s.next = c.SampleFrom
+	s.networkHead = c.NetworkHead
+	// put failed into priority to retry them on restart
+	for h, count := range c.Failed {
+		s.failed[h] = count
+		s.priority = append(s.priority, s.newJob(h, h))
+	}
+}
+
+func (s *coordinatorState) handleResult(res result) {
+	delete(s.inProgress, res.id)
+
+	failedFromWorker := make(map[uint64]bool)
+	for _, h := range res.failed {
+		failedFromWorker[h] = true
+	}
+
+	// check if the worker retried any of the previously failed heights
+	for h := range s.failed {
+		if h < res.From || h > res.To {
+			continue
+		}
+
+		if !failedFromWorker[h] {
+			delete(s.failed, h)
+		}
+	}
+	// add newly failed heights
+	for h := range failedFromWorker {
+		s.failed[h]++
+	}
+	s.checkDone()
 }
 
-// JobInfo contains information about a catchUp job.
-type JobInfo struct {
-	Start time.Time `json:"start"`
-	End   time.Time `json:"end"`
-	Error error     `json:"error"`
+func (s *coordinatorState) updateHead(last uint64) bool {
```

Comment:
why does this return a bool if it's never checked?

Reply:
I forgot to remove it after branching metrics into separate PR. It will be used in metrics PR to trigger new_head metric.